libkbfs/mdserver_disk.go

```diff
@@ -5,97 +5,66 @@
 package libkbfs
 
 import (
-	"bytes"
-	"encoding/binary"
 	"errors"
-	"fmt"
 	"io/ioutil"
 	"os"
 	"path/filepath"
 	"reflect"
 	"sync"
-	"time"
 
 	"github.com/keybase/client/go/logger"
 	keybase1 "github.com/keybase/client/go/protocol"
 	"github.com/syndtr/goleveldb/leveldb"
 	"github.com/syndtr/goleveldb/leveldb/storage"
-	"github.com/syndtr/goleveldb/leveldb/util"
 	"golang.org/x/net/context"
 )
 
-// TODO: Convert this to use flat files and a journal.
+type mdServerDiskShared struct {
+	dirPath string
 
-// MDServerDisk just stores blocks in local leveldb instances.
-type MDServerDisk struct {
-	config   Config
-	handleDb *leveldb.DB // folder handle                  -> folderId
-	mdDb     *leveldb.DB // folderId+[branchId]+[revision] -> mdBlockLocal
-	branchDb *leveldb.DB // folderId+deviceKID             -> branchId
-	log      logger.Logger
-
-	locksMutex *sync.Mutex
-	locksDb    *leveldb.DB // folderId -> deviceKID
+	// Protects handleDb and tlfStorage. After Shutdown() is
+	// called, both handleDb and tlfStorage are nil.
+	lock       sync.RWMutex
+	handleDb   *leveldb.DB // folder handle -> folderId
+	tlfStorage map[TlfID]*mdServerTlfStorage
 
 	updateManager *mdServerLocalUpdateManager
 
-	shutdownLock *sync.RWMutex
-	shutdown     *bool
 	shutdownFunc func(logger.Logger)
 }
 
-var _ mdServerLocal = (*MDServerDisk)(nil)
+// MDServerDisk stores all info on disk.
+type MDServerDisk struct {
+	codec  Codec
+	clock  Clock
+	crypto Crypto
+	kbpki  KBPKI
+	log    logger.Logger
 
-type mdBlockLocal struct {
-	MD        *RootMetadataSigned
-	Timestamp time.Time
+	*mdServerDiskShared
 }
 
+var _ mdServerLocal = (*MDServerDisk)(nil)
+
 func newMDServerDisk(config Config, dirPath string,
 	shutdownFunc func(logger.Logger)) (*MDServerDisk, error) {
 	handlePath := filepath.Join(dirPath, "handles")
-	mdPath := filepath.Join(dirPath, "md")
-	branchPath := filepath.Join(dirPath, "branches")
 
 	handleStorage, err := storage.OpenFile(handlePath)
 	if err != nil {
 		return nil, err
 	}
 
-	mdStorage, err := storage.OpenFile(mdPath)
-	if err != nil {
-		return nil, err
-	}
-
-	branchStorage, err := storage.OpenFile(branchPath)
-	if err != nil {
-		return nil, err
-	}
-
-	// Always use memory for the lock storage, so it gets wiped after
-	// a restart.
-	lockStorage := storage.NewMemStorage()
-
 	handleDb, err := leveldb.Open(handleStorage, leveldbOptions)
 	if err != nil {
 		return nil, err
 	}
-	mdDb, err := leveldb.Open(mdStorage, leveldbOptions)
-	if err != nil {
-		return nil, err
-	}
-	branchDb, err := leveldb.Open(branchStorage, leveldbOptions)
-	if err != nil {
-		return nil, err
-	}
-	locksDb, err := leveldb.Open(lockStorage, leveldbOptions)
-	if err != nil {
-		return nil, err
-	}
 	log := config.MakeLogger("")
-	mdserv := &MDServerDisk{config, handleDb, mdDb, branchDb, log,
-		&sync.Mutex{}, locksDb, newMDServerLocalUpdateManager(),
-		&sync.RWMutex{}, new(bool), shutdownFunc}
+	mdserv := &MDServerDisk{config.Codec(), config.Clock(),
```

Comment:
Slight preference for using field names in the constructor here, to avoid having to do `sync.RWMutex{}`.